crates/astria-core/src/primitive/v1/asset/denom.rs

```diff
@@ -0,0 +1,554 @@
+use std::{
+    collections::VecDeque,
+    str::FromStr,
+};
+
+/// Represents a denomination of a sequencer asset.
+///
+/// This can be either an IBC-bridged asset or a native asset.
+/// If it's a native asset, the prefix will be empty.
+///
+/// Note that the full denomination trace of the token is `prefix/base_denom`,
+/// in the case that a prefix is present.
+/// This is hashed to create the ID.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Denom {
+    TracePrefixed(TracePrefixed),
+    IbcPrefixed(IbcPrefixed),
+}
+
+impl Denom {
+    #[must_use]
+    pub fn is_ibc_prefixed(&self) -> bool {
+        self.as_ibc_prefixed().is_some()
+    }
+
+    #[must_use]
+    pub fn is_trace_prefixed(&self) -> bool {
+        self.as_trace_prefixed().is_some()
+    }
+
+    #[must_use]
+    pub fn id(&self) -> super::Id {
+        match self {
+            Self::TracePrefixed(trace) => trace.id(),
+            Self::IbcPrefixed(ibc) => ibc.id(),
+        }
+    }
+
+    #[must_use]
+    pub fn as_ibc_prefixed(&self) -> Option<&IbcPrefixed> {
+        match self {
+            Denom::TracePrefixed(_) => None,
+            Denom::IbcPrefixed(ibc_prefixed) => Some(ibc_prefixed),
+        }
+    }
+
+    #[must_use]
+    pub fn as_trace_prefixed(&self) -> Option<&TracePrefixed> {
+        match self {
+            Denom::TracePrefixed(trace) => Some(trace),
+            Denom::IbcPrefixed(_) => None,
+        }
+    }
+
+    /// Unwraps the inner ibc prefixed denom.
+    ///
+    /// # Panics
+    /// Panics if the self value equals [`Self::TracePrefixed`].
+    #[must_use]
+    pub fn unwrap_ibc_prefixed(self) -> IbcPrefixed {
+        let Self::IbcPrefixed(ibc) = self else {
+            panic!("not ibc");
+        };
+        ibc
+    }
+
+    /// Unwraps the inner trace prefixed denom.
+    ///
+    /// # Panics
+    /// Panics if the self value equals [`Self::IbcPrefixed`].
+    #[must_use]
+    pub fn unwrap_trace(self) -> TracePrefixed {
```

Comment:
```suggestion
    pub fn unwrap_trace_prefix(self) -> TracePrefixed {
```

for consistency w above

Reply:
Fixed in 7b8ce7f (but used `unwrap_trace_prefixed`)